File: python/src/revenant/core/pdf/cms_signature.py
# SPDX-License-Identifier: Apache-2.0
# pyright: reportUnknownMemberType=false, reportUnknownVariableType=false, reportUnknownArgumentType=false, reportIndexIssue=false, reportAttributeAccessIssue=false
"""Cryptographic verification of CMS ``SignerInfo`` signatures.

The PDF verifier checks two independent properties: the CMS ``messageDigest``
must match the document's ByteRange, and the signature over the DER-encoded
signed attributes must verify with the certificate identified by ``SignerInfo``.
This module implements the second check and deliberately fails closed whenever
the CMS cannot be verified.
"""

from __future__ import annotations

import hmac
import logging
from dataclasses import dataclass
from typing import ClassVar, Literal

from asn1crypto import algos as asn1_algos
from asn1crypto import cms as asn1_cms
from asn1crypto import core as asn1_core
from cryptography.exceptions import InvalidSignature, UnsupportedAlgorithm
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import padding, rsa
from cryptography.hazmat.primitives.serialization import load_der_public_key

from ..cms_certificates import find_signer_certificate
from .cms_info import resolve_hash_algo

_logger = logging.getLogger(__name__)

_OID_SIGNED_DATA = "1.2.840.113549.1.7.2"
_OID_CONTENT_TYPE = "1.2.840.113549.1.9.3"
_OID_MESSAGE_DIGEST = "1.2.840.113549.1.9.4"
_TAG_CONTEXT_0_CONSTRUCTED = 0xA0  # [0] IMPLICIT, as signed attrs appear in SignerInfo
_TAG_SET_OF = 0x31  # universal SET OF, as required for the signature input

# Diagnostics are surfaced to users through VerificationResult.details.
_MAX_REASON_LENGTH = 160

_OID_SIGNING_CERTIFICATE = "1.2.840.113549.1.9.16.2.12"
_OID_SIGNING_CERTIFICATE_V2 = "1.2.840.113549.1.9.16.2.47"

_RSA_SIGNATURE_DIGESTS: dict[str, str | None] = {
    "1.2.840.113549.1.1.1": None,  # rsaEncryption (hash is in digestAlgorithm)
    "1.2.840.113549.1.1.5": "sha1",  # sha1WithRSAEncryption
    "1.2.840.113549.1.1.11": "sha256",  # sha256WithRSAEncryption
    "1.2.840.113549.1.1.12": "sha384",  # sha384WithRSAEncryption
    "1.2.840.113549.1.1.13": "sha512",  # sha512WithRSAEncryption
}

_HASH_ALGORITHMS: dict[str, type[hashes.HashAlgorithm]] = {
    "sha1": hashes.SHA1,
    "sha256": hashes.SHA256,
    "sha384": hashes.SHA384,
    "sha512": hashes.SHA512,
}


class _ESSCertID(asn1_core.Sequence):
    _fields: ClassVar[list[object]] = [
        ("cert_hash", asn1_core.OctetString),
        ("issuer_serial", asn1_core.Any, {"optional": True}),
    ]


class _ESSCertIDs(asn1_core.SequenceOf):
    _child_spec = _ESSCertID


class _SigningCertificate(asn1_core.Sequence):
    _fields: ClassVar[list[object]] = [
        ("certs", _ESSCertIDs),
        ("policies", asn1_core.Any, {"optional": True}),
    ]


class _ESSCertIDv2(asn1_core.Sequence):
    _fields: ClassVar[list[object]] = [
        (
            "hash_algorithm",
            asn1_algos.DigestAlgorithm,
            {"default": {"algorithm": "sha256"}},
        ),
        ("cert_hash", asn1_core.OctetString),
        ("issuer_serial", asn1_core.Any, {"optional": True}),
    ]


class _ESSCertIDsV2(asn1_core.SequenceOf):
    _child_spec = _ESSCertIDv2


class _SigningCertificateV2(asn1_core.Sequence):
    _fields: ClassVar[list[object]] = [
        ("certs", _ESSCertIDsV2),
        ("policies", asn1_core.Any, {"optional": True}),
    ]


@dataclass(frozen=True)
class SignatureVerification:
    """Tri-state CMS signature result.

    ``valid`` is ``None`` when verification could not be performed (for
    example, because the signer certificate or algorithm is unsupported).
    Both ``False`` and ``None`` fail the overall verification verdict.
    """

    valid: bool | None
    reason: str | None = None
    signer_certificate_bound: bool = False
    #: True when the CMS carries no signed attributes and the signature was
    #: verified over the caller-supplied content itself, so no separate digest
    #: exists to compare and the signature verdict is the integrity check.
    #: Never true for content the caller did not supply.
    covers_content: bool = False

    def describe(self) -> str:
        """Return a stable, human-readable diagnostic line."""
        if self.valid is True:
            return "Signature OK -- signer signature verifies"
        if self.valid is False:
            return "Signature INVALID -- does not verify against the signer certificate"
        reason = self.reason or "verification unavailable"
        return f"Signature not verified ({reason})"


def _unverifiable(reason: str) -> SignatureVerification:
    return SignatureVerification(valid=None, reason=reason)


def _describe_exception(error: Exception) -> str:
    """Render an exception compactly enough to sit in a user-facing detail line."""
    message = str(error).strip().replace("\n", " ")
    if len(message) > _MAX_REASON_LENGTH:
        message = message[: _MAX_REASON_LENGTH - 3] + "..."
    return f"{type(error).__name__}: {message}" if message else type(error).__name__


def _resolve_digest_algorithm(signer_info: asn1_cms.SignerInfo) -> hashes.HashAlgorithm | None:
    algo_id = signer_info["digest_algorithm"]["algorithm"]
    algo_name = resolve_hash_algo(algo_id.native)
    if algo_name is None:
        algo_name = resolve_hash_algo(algo_id.dotted)
    hash_type = _HASH_ALGORITHMS.get(algo_name) if algo_name is not None else None
    return hash_type() if hash_type is not None else None


def _validate_signed_attributes(
    signed_data: asn1_cms.SignedData,
    signer_info: asn1_cms.SignerInfo,
    digest_algorithm: hashes.HashAlgorithm,
) -> str | None:
    """Validate mandatory signed attributes and their cardinality."""
    signed_attrs = signer_info["signed_attrs"]
    if signed_attrs.native is None:
        return "no signed attributes"

    content_type_attrs = [attr for attr in signed_attrs if attr["type"].dotted == _OID_CONTENT_TYPE]
    if len(content_type_attrs) != 1 or len(content_type_attrs[0]["values"]) != 1:
        return "contentType attribute missing or duplicated"

    try:
        signed_content_type = content_type_attrs[0]["values"][0].dotted
        encapsulated_content_type = signed_data["encap_content_info"]["content_type"].dotted
    except (AttributeError, KeyError, TypeError, IndexError):
        return "contentType attribute is malformed"
    if signed_content_type != encapsulated_content_type:
        return "contentType attribute is inconsistent"

    digest_attrs = [attr for attr in signed_attrs if attr["type"].dotted == _OID_MESSAGE_DIGEST]
    if len(digest_attrs) != 1 or len(digest_attrs[0]["values"]) != 1:
        return "messageDigest attribute missing or duplicated"
    digest_value = digest_attrs[0]["values"][0].native
    if not isinstance(digest_value, bytes) or len(digest_value) != digest_algorithm.digest_size:
        return "messageDigest attribute is malformed"

    return None


def _retag_as_set_of(encoded: bytes) -> bytes | None:
    """Replace the ``[0] IMPLICIT`` tag with the universal SET OF tag."""
    if not encoded:
        return None
    retagged = bytearray(encoded)
    if retagged[0] == _TAG_CONTEXT_0_CONSTRUCTED:
        retagged[0] = _TAG_SET_OF
    elif retagged[0] != _TAG_SET_OF:
        return None
    return bytes(retagged)


def _signed_attributes_signature_inputs(signer_info: asn1_cms.SignerInfo) -> list[bytes]:
    """Return the candidate byte strings the signer may have signed.

    RFC 5652 section 5.4 defines the signature input as the DER encoding of the
    signed attributes with the ``[0] IMPLICIT`` tag replaced by a universal
    ``SET OF`` tag.  Signers that emit a non-DER attribute ordering nevertheless
    sign the bytes they transmitted, so the as-transmitted encoding is tried
    first and the canonical DER re-encoding second.

    Accepting either encoding does not weaken the check: both encode the very
    same parsed attributes, and those attributes are validated independently by
    :func:`_validate_signed_attributes` before any signature is checked.
    """
    signed_attrs = signer_info["signed_attrs"]
    candidates: list[bytes] = []
    for encoded in (signed_attrs.dump(), signed_attrs.dump(force=True)):
        retagged = _retag_as_set_of(encoded)
        if retagged is not None and retagged not in candidates:
            candidates.append(retagged)
    return candidates


SigningCertificateBinding = Literal["absent", "match", "mismatch", "unparsable"]


def _single_signing_certificate_binding(
    binding_attr: asn1_cms.CMSAttribute,
    signer_cert_der: bytes,
) -> SigningCertificateBinding:
    """Evaluate one ESS certificate-binding attribute."""
    if len(binding_attr["values"]) != 1:
        return "unparsable"

    is_v2 = binding_attr["type"].dotted == _OID_SIGNING_CERTIFICATE_V2
    try:
        value_der = binding_attr["values"][0].dump()
        if is_v2:
            parsed = _SigningCertificateV2.load(value_der, strict=True)
            if not parsed["certs"]:
                return "unparsable"
            cert_id = parsed["certs"][0]
            algorithm_id = cert_id["hash_algorithm"]["algorithm"]
            algorithm_name = resolve_hash_algo(algorithm_id.native)
            if algorithm_name is None:
                algorithm_name = resolve_hash_algo(algorithm_id.dotted)
            hash_type = _HASH_ALGORITHMS.get(algorithm_name) if algorithm_name else None
            if hash_type is None:
                return "unparsable"
            hash_algorithm = hash_type()
        else:
            parsed = _SigningCertificate.load(value_der, strict=True)
            if not parsed["certs"]:
                return "unparsable"
            cert_id = parsed["certs"][0]
            # RFC 2634 fixes ESSCertID v1 to SHA-1, so there is no stronger
            # option inside this attribute, and refusing v1 outright is not open
            # to us either: real signers still emit it alone. It is checked
            # because checking beats ignoring, not because SHA-1 is sound -- a
            # certificate-hash binding does rest on collision resistance, so it
            # counts as defence in depth over chain validation rather than proof
            # by itself. Every binding present must match (see
            # _signing_certificate_binding), so a signer that also emits v2
            # cannot be satisfied by a v1 collision alone.
            hash_algorithm = hashes.SHA1()  # noqa: S303

        expected_hash = cert_id["cert_hash"].native
        if not isinstance(expected_hash, bytes):
            return "unparsable"
    except Exception:
        return "unparsable"

    digest = hashes.Hash(hash_algorithm)
    digest.update(signer_cert_der)
    if not hmac.compare_digest(digest.finalize(), expected_hash):
        return "mismatch"
    return "match"


def _signing_certificate_binding(
    signer_info: asn1_cms.SignerInfo,
    signer_cert_der: bytes,
) -> SigningCertificateBinding:
    """Classify all optional ESS certificate bindings against the signer cert."""
    signed_attrs = signer_info["signed_attrs"]
    binding_attrs = [
        attr
        for attr in signed_attrs
        if attr["type"].dotted in {_OID_SIGNING_CERTIFICATE, _OID_SIGNING_CERTIFICATE_V2}
    ]
    if not binding_attrs:
        return "absent"

    binding_oids = [attr["type"].dotted for attr in binding_attrs]
    if len(set(binding_oids)) != len(binding_oids):
        return "unparsable"

    statuses = [
        _single_signing_certificate_binding(binding_attr, signer_cert_der)
        for binding_attr in binding_attrs
    ]
    if "mismatch" in statuses:
        return "mismatch"
    if "unparsable" in statuses:
        return "unparsable"
    return "match"


def verify_signer_signature(
    cms_der: bytes,
    content: bytes | None = None,
) -> SignatureVerification:
    """Verify the first CMS signer's RSA PKCS#1 v1.5 signature.

    Certificate-chain trust is intentionally not checked here; the caller
    reports it separately. Unsupported algorithms, missing certificates, and
    malformed CMS values are reported as unverifiable and never accepted.

    Args:
        cms_der: DER-encoded CMS/PKCS#7 ``SignedData``.
        content: The detached content the signature covers. Consulted when the
            CMS carries no signed attributes -- RFC 5652 section 5.4 defines the
            signature input as the content itself in that case, and signers do
            issue documents in that shape. A CMS that also embeds its own content
            is rejected unless the two are byte-identical.
    """
    try:
        content_info = asn1_cms.ContentInfo.load(cms_der, strict=True)
        if content_info["content_type"].dotted != _OID_SIGNED_DATA:
            return _unverifiable("CMS content is not SignedData")

        signed_data = content_info["content"]
        signer_infos = signed_data["signer_infos"]
        if not signer_infos:
            return _unverifiable("no SignerInfo present")
        signer_info = signer_infos[0]

        digest_algorithm = _resolve_digest_algorithm(signer_info)
        if digest_algorithm is None:
            return _unverifiable("unrecognized digest algorithm")

        signature_oid = signer_info["signature_algorithm"]["algorithm"].dotted
        if signature_oid not in _RSA_SIGNATURE_DIGESTS:
            return _unverifiable("non-RSA signer is unsupported")
        signature_digest = _RSA_SIGNATURE_DIGESTS[signature_oid]
        if signature_digest is not None and signature_digest != digest_algorithm.name:
            return _unverifiable("signatureAlgorithm conflicts with digestAlgorithm")

        # RFC 5652 section 5.4: the signature covers the DER-encoded signed
        # attributes when they are present, and the content itself when they are
        # not. The branch is taken strictly on presence -- a CMS that carries
        # signed attributes can never fall back to the content path, so a
        # malformed attribute set still fails closed.
        signed_attrs_present = signer_info["signed_attrs"].native is not None
        if signed_attrs_present:
            attr_error = _validate_signed_attributes(signed_data, signer_info, digest_algorithm)
            if attr_error is not None:
                return _unverifiable(attr_error)
            signature_inputs = _signed_attributes_signature_inputs(signer_info)
            if not signature_inputs:
                return _unverifiable("cannot encode signed attributes")
        else:
            # RFC 5652 section 5.4 makes the content itself the signature input,
            # which leaves the question of whose content. A CMS may carry its own
            # eContent, and a signature over those bytes says nothing about the
            # caller's. Preferring the embedded copy silently re-attributes the
            # verdict to data the signer never saw, so the caller's bytes win and
            # an embedded copy is only ever a duplicate of them.
            embedded = signed_data["encap_content_info"]["content"].native
            if content is not None and embedded is not None and embedded != content:
                return _unverifiable("embedded CMS content differs from the data being verified")
            body = content if content is not None else embedded
            if body is None:
                return _unverifiable("no signed attributes and no content to verify")
            signature_inputs = [body]

        # Without signed attributes there is no messageDigest for the caller to
        # re-check, so this flag is the whole integrity argument downstream. It
        # may only be raised when the bytes fed to the signature were the ones
        # the caller asked about.
        covers_caller_content = not signed_attrs_present and content is not None

        signer_cert = find_signer_certificate(signed_data, signer_info)
        if signer_cert is None:
            return _unverifiable("signer certificate not embedded or ambiguous")

        signer_cert_der = signer_cert.dump()
        # Only the public key is needed, so load the SubjectPublicKeyInfo rather
        # than the whole certificate. Signer certificates issued in the field
        # encode DN attributes such as emailAddress as a PrintableString
        # holding '@', which is outside that type's character set; a strict X.509
        # parser rejects the entire certificate over a field this check never
        # reads. Signature verification must not depend on decoding the subject DN.
        try:
            public_key = load_der_public_key(
                signer_cert["tbs_certificate"]["subject_public_key_info"].dump()
            )
        except (ValueError, UnsupportedAlgorithm):
            return _unverifiable("signer public key is unusable")
        if not isinstance(public_key, rsa.RSAPublicKey):
            return _unverifiable("signer public key is not RSA")

        signature = signer_info["signature"].native
        for signature_input in signature_inputs:
            try:
                public_key.verify(
                    signature,
                    signature_input,
                    padding.PKCS1v15(),
                    digest_algorithm,
                )
                break
            except InvalidSignature:
                continue
        else:
            return SignatureVerification(valid=False, covers_content=covers_caller_content)

        # An ESS binding lives in the signed attributes, so without them there is
        # nothing to bind the certificate beyond the signature itself.
        binding: SigningCertificateBinding = "absent"
        if signed_attrs_present:
            binding = _signing_certificate_binding(signer_info, signer_cert_der)
            if binding == "mismatch":
                return _unverifiable("signingCertificate attribute names a different certificate")
            if binding == "unparsable":
                return _unverifiable("signingCertificate attribute could not be parsed")
    except Exception as e:
        # CMS is untrusted input and this API promises a verdict rather than a
        # parser exception, so it fails closed.  The reason names the underlying
        # exception: an opaque "check failed" once masked a certificate-parsing
        # incompatibility that silently invalidated genuine signatures, and a
        # diagnostic nobody can read is how that reaches production unnoticed.
        _logger.warning("Could not verify CMS signer signature: %s", e, exc_info=True)
        return _unverifiable(f"CMS signature check failed -- {_describe_exception(e)}")

    return SignatureVerification(
        valid=True,
        signer_certificate_bound=binding == "match",
        covers_content=covers_caller_content,
    )

File: typescript/src/core/pdf/cms-signature.ts
// SPDX-License-Identifier: Apache-2.0
/** Cryptographic verification of CMS SignerInfo signatures. */

import { createHash } from "node:crypto";

import * as asn1js from "asn1js";
import * as pkijs from "pkijs";

import { toArrayBuffer } from "../../utils.js";
import { findSignerCertificate } from "../cms-certificates.js";
import { resolveHashAlgo } from "./cms-info.js";

const OID_SIGNED_DATA = "1.2.840.113549.1.7.2";
const OID_CONTENT_TYPE = "1.2.840.113549.1.9.3";
const OID_MESSAGE_DIGEST = "1.2.840.113549.1.9.4";
const OID_SIGNING_CERTIFICATE = "1.2.840.113549.1.9.16.2.12";
const OID_SIGNING_CERTIFICATE_V2 = "1.2.840.113549.1.9.16.2.47";
const OID_RSA_ENCRYPTION = "1.2.840.113549.1.1.1";

const RSA_SIGNATURE_DIGESTS: ReadonlyMap<string, string | null> = new Map([
  ["1.2.840.113549.1.1.1", null], // rsaEncryption: digestAlgorithm is authoritative
  ["1.2.840.113549.1.1.5", "SHA-1"],
  ["1.2.840.113549.1.1.11", "SHA-256"],
  ["1.2.840.113549.1.1.12", "SHA-384"],
  ["1.2.840.113549.1.1.13", "SHA-512"],
]);

const DIGEST_LENGTHS: ReadonlyMap<string, number> = new Map([
  ["SHA-1", 20],
  ["SHA-256", 32],
  ["SHA-384", 48],
  ["SHA-512", 64],
]);

export interface SignatureVerification {
  /** true = verified, false = bad signature, null = verification unavailable. */
  valid: boolean | null;
  /** Whether a matching signed ESS attribute binds the exact signer certificate. */
  signerCertificateBound: boolean;
  /**
   * True when the CMS carries no signed attributes, so the signature was
   * verified over the content itself and no separate digest exists to compare.
   */
  coversContent: boolean;
  /** Stable diagnostic line for VerificationResult.details. */
  detail: string;
}

/** Diagnostics are surfaced to users through VerificationResult.details. */
const MAX_REASON_LENGTH = 160;

/** Render a thrown value compactly enough to sit in a user-facing detail line. */
export function describeError(error: unknown): string {
  const name = error instanceof Error ? error.name : "Error";
  const raw = error instanceof Error ? error.message : String(error);
  const message = raw.trim().replace(/\s+/g, " ");
  if (message.length === 0) return name;
  const trimmed =
    message.length > MAX_REASON_LENGTH ? `${message.slice(0, MAX_REASON_LENGTH - 3)}...` : message;
  return `${name}: ${trimmed}`;
}

/** DER SET OF ordering: components sort by their complete encodings (X.690 11.6). */
function compareDerElements(a: Uint8Array, b: Uint8Array): number {
  const shared = Math.min(a.length, b.length);
  for (let i = 0; i < shared; i++) {
    const left = a[i] ?? 0;
    const right = b[i] ?? 0;
    if (left !== right) return left - right;
  }
  return a.length - b.length;
}

function derLength(size: number): Uint8Array {
  if (size < 0x80) return new Uint8Array([size]);
  const octets: number[] = [];
  for (let rest = size; rest > 0; rest = Math.floor(rest / 256)) octets.unshift(rest % 256);
  return new Uint8Array([0x80 | octets.length, ...octets]);
}

/**
 * Re-encode the signed attributes as a canonical DER SET OF.
 *
 * RFC 5652 section 5.4 defines the signature input this way, and a conforming
 * signer transmits exactly these bytes. Signers that emit another ordering sign
 * what they transmitted instead, so both encodings are tried; they encode the
 * same parsed attributes, which validateSignedAttributes has already checked.
 */
function canonicalSignedAttributes(signedAttrs: pkijs.SignedAndUnsignedAttributes): Uint8Array {
  const members = signedAttrs.attributes.map(
    (attribute) => new Uint8Array(attribute.toSchema().toBER(false)),
  );
  members.sort(compareDerElements);
  const body = new Uint8Array(members.reduce((total, member) => total + member.length, 0));
  let offset = 0;
  for (const member of members) {
    body.set(member, offset);
    offset += member.length;
  }
  const header = derLength(body.length);
  const encoded = new Uint8Array(1 + header.length + body.length);
  encoded[0] = 0x31; // universal SET OF, as required for the signature input
  encoded.set(header, 1);
  encoded.set(body, 1 + header.length);
  return encoded;
}

function unverifiable(reason: string): SignatureVerification {
  return {
    valid: null,
    signerCertificateBound: false,
    coversContent: false,
    detail: `Signature not verified (${reason})`,
  };
}

function validateSignedAttributes(
  signedData: pkijs.SignedData,
  signerInfo: pkijs.SignerInfo,
  digestAlgorithm: string,
): string | null {
  const signedAttrs = signerInfo.signedAttrs;
  if (!signedAttrs) return "no signed attributes";

  const contentTypeAttrs = signedAttrs.attributes.filter((attr) => attr.type === OID_CONTENT_TYPE);
  if (contentTypeAttrs.length !== 1 || contentTypeAttrs[0]?.values.length !== 1) {
    return "contentType attribute missing or duplicated";
  }
  const contentTypeValue = contentTypeAttrs[0].values[0];
  if (!(contentTypeValue instanceof asn1js.ObjectIdentifier)) {
    return "contentType attribute is malformed";
  }
  if (contentTypeValue.valueBlock.toString() !== signedData.encapContentInfo.eContentType) {
    return "contentType attribute is inconsistent";
  }

  const digestAttrs = signedAttrs.attributes.filter((attr) => attr.type === OID_MESSAGE_DIGEST);
  if (digestAttrs.length !== 1 || digestAttrs[0]?.values.length !== 1) {
    return "messageDigest attribute missing or duplicated";
  }
  const digestValue = digestAttrs[0].values[0];
  const expectedLength = DIGEST_LENGTHS.get(digestAlgorithm);
  if (
    !(digestValue instanceof asn1js.OctetString) ||
    expectedLength === undefined ||
    digestValue.valueBlock.valueHexView.byteLength !== expectedLength
  ) {
    return "messageDigest attribute is malformed";
  }

  return null;
}

interface EssCertHash {
  algorithm: string;
  value: Uint8Array;
}

type SigningCertificateBinding = "absent" | "match" | "mismatch" | "unparsable";

function bytesEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}

function parseEssCertHash(value: asn1js.BaseBlock, isV2: boolean): EssCertHash | null {
  if (!(value instanceof asn1js.Sequence)) return null;
  const certs = value.valueBlock.value[0];
  if (!(certs instanceof asn1js.Sequence)) return null;
  const firstCertId = certs.valueBlock.value[0];
  if (!(firstCertId instanceof asn1js.Sequence)) return null;

  const fields = firstCertId.valueBlock.value;
  let algorithm = "SHA-1";
  let hashIndex = 0;
  if (isV2) {
    algorithm = "SHA-256";
    const optionalAlgorithm = fields[0];
    if (optionalAlgorithm instanceof asn1js.Sequence) {
      const algorithmIdentifier = new pkijs.AlgorithmIdentifier({ schema: optionalAlgorithm });
      const resolved = resolveHashAlgo(algorithmIdentifier.algorithmId);
      if (!resolved) return null;
      algorithm = resolved;
      hashIndex = 1;
    }
  }

  const certHash = fields[hashIndex];
  const expectedLength = DIGEST_LENGTHS.get(algorithm);
  if (
    !(certHash instanceof asn1js.OctetString) ||
    expectedLength === undefined ||
    certHash.valueBlock.valueHexView.byteLength !== expectedLength
  ) {
    return null;
  }
  return {
    algorithm,
    value: new Uint8Array(certHash.valueBlock.valueHexView),
  };
}

function signingCertificateBinding(
  signerInfo: pkijs.SignerInfo,
  signerCertificate: pkijs.Certificate,
): SigningCertificateBinding {
  const bindingAttrs = (signerInfo.signedAttrs?.attributes ?? []).filter(
    (attribute) =>
      attribute.type === OID_SIGNING_CERTIFICATE || attribute.type === OID_SIGNING_CERTIFICATE_V2,
  );
  if (bindingAttrs.length === 0) return "absent";
  if (new Set(bindingAttrs.map((attribute) => attribute.type)).size !== bindingAttrs.length) {
    return "unparsable";
  }

  let certificateDer: Uint8Array;
  try {
    certificateDer = new Uint8Array(signerCertificate.toSchema().toBER(false));
  } catch {
    return "unparsable";
  }
  let unparsable = false;
  for (const bindingAttr of bindingAttrs) {
    try {
      if (bindingAttr.values.length !== 1) {
        unparsable = true;
        continue;
      }
      const bindingValue = bindingAttr.values[0];
      if (!bindingValue) {
        unparsable = true;
        continue;
      }
      const parsed = parseEssCertHash(
        bindingValue,
        bindingAttr.type === OID_SIGNING_CERTIFICATE_V2,
      );
      if (!parsed) {
        unparsable = true;
        continue;
      }

      const nodeAlgorithm = parsed.algorithm.toLowerCase().replace("-", "");
      const actualHash = new Uint8Array(createHash(nodeAlgorithm).update(certificateDer).digest());
      if (!bytesEqual(actualHash, parsed.value)) return "mismatch";
    } catch {
      unparsable = true;
    }
  }

  return unparsable ? "unparsable" : "match";
}

/**
 * Verify the first CMS signer's signature with its embedded certificate.
 *
 * Certificate-chain trust is deliberately separate. Unsupported or malformed
 * inputs fail closed and can never produce a valid overall verification result.
 */
export async function verifySignerSignature(
  cmsDer: Uint8Array,
  content: Uint8Array | null = null,
): Promise<SignatureVerification> {
  try {
    const asn1 = asn1js.fromBER(toArrayBuffer(cmsDer));
    if (asn1.offset === -1 || asn1.offset !== cmsDer.byteLength) {
      return unverifiable("CMS did not parse as one complete ASN.1 value");
    }

    const contentInfo = new pkijs.ContentInfo({ schema: asn1.result });
    if (contentInfo.contentType !== OID_SIGNED_DATA) {
      return unverifiable("CMS content is not SignedData");
    }

    const signedData = new pkijs.SignedData({ schema: contentInfo.content });
    const signerInfo = signedData.signerInfos[0];
    if (!signerInfo) return unverifiable("no SignerInfo present");
    const signerCertificate = findSignerCertificate(signedData, signerInfo);
    if (signerCertificate === null) {
      return unverifiable("signer certificate not embedded or ambiguous");
    }

    const digestAlgorithm = resolveHashAlgo(signerInfo.digestAlgorithm.algorithmId);
    if (!digestAlgorithm) return unverifiable("unrecognized digest algorithm");

    const signatureAlgorithm = signerInfo.signatureAlgorithm.algorithmId;
    if (!RSA_SIGNATURE_DIGESTS.has(signatureAlgorithm)) {
      return unverifiable("non-RSA signer is unsupported");
    }
    const signatureDigest = RSA_SIGNATURE_DIGESTS.get(signatureAlgorithm);
    if (signatureDigest !== null && signatureDigest !== digestAlgorithm) {
      return unverifiable("signatureAlgorithm conflicts with digestAlgorithm");
    }
    if (signerCertificate.subjectPublicKeyInfo.algorithm.algorithmId !== OID_RSA_ENCRYPTION) {
      return unverifiable("RSA signatureAlgorithm used with a non-RSA signer key");
    }

    // RFC 5652 section 5.4: the signature covers the DER-encoded signed
    // attributes when they are present, and the content itself when they are
    // not. The branch is taken strictly on presence, so a CMS that carries
    // signed attributes can never fall back to the content path.
    const signedAttrs = signerInfo.signedAttrs;
    let signatureInputs: ArrayBuffer[];
    if (signedAttrs) {
      const attrError = validateSignedAttributes(signedData, signerInfo, digestAlgorithm);
      if (attrError) return unverifiable(attrError);
      if (signedAttrs.encodedValue.byteLength === 0) {
        return unverifiable("cannot encode signed attributes");
      }
      // The as-transmitted encoding first, then the canonical DER re-encoding.
      const transmitted = new Uint8Array(signedAttrs.encodedValue);
      const canonical = canonicalSignedAttributes(signedAttrs);
      signatureInputs = bytesEqual(transmitted, canonical)
        ? [signedAttrs.encodedValue]
        : [signedAttrs.encodedValue, toArrayBuffer(canonical)];
    } else {
      const eContent = signedData.encapContentInfo.eContent;
      const body = eContent ? new Uint8Array(eContent.valueBlock.valueHexView) : content;
      if (!body) return unverifiable("no signed attributes and no content to verify");
      signatureInputs = [toArrayBuffer(body)];
    }

    // SignedData.verify() performs its own signer-certificate lookup. For an
    // SKI SignerIdentifier, PKIjs 3.4.0 hashes SubjectPublicKey instead of
    // matching the certificate's Subject Key Identifier extension, so it can
    // verify with a different certificate than findSignerCertificate selected.
    // Invoke the crypto primitive directly with the already selected key.
    const crypto = pkijs.getCrypto(true);
    let verified = false;
    for (const signatureInput of signatureInputs) {
      verified = await crypto.verifyWithPublicKey(
        signatureInput,
        signerInfo.signature,
        signerCertificate.subjectPublicKeyInfo,
        signerInfo.signatureAlgorithm,
        digestAlgorithm,
      );
      if (verified) break;
    }
    if (!verified) {
      return {
        valid: false,
        signerCertificateBound: false,
        coversContent: !signedAttrs,
        detail: "Signature INVALID -- does not verify against the signer certificate",
      };
    }

    // An ESS binding lives in the signed attributes, so without them there is
    // nothing to bind the certificate beyond the signature itself.
    let binding: SigningCertificateBinding = "absent";
    if (signedAttrs) {
      binding = signingCertificateBinding(signerInfo, signerCertificate);
      if (binding === "mismatch") {
        return unverifiable("signingCertificate attribute names a different certificate");
      }
      if (binding === "unparsable") {
        return unverifiable("signingCertificate attribute could not be parsed");
      }
    }
    return {
      valid: true,
      signerCertificateBound: binding === "match",
      coversContent: !signedAttrs,
      detail: "Signature OK -- signer signature verifies",
    };
  } catch (error) {
    // CMS is untrusted input and this API promises a verdict rather than a
    // thrown parser error, so it fails closed. The reason names the underlying
    // failure: an opaque "check failed" once masked a certificate-parsing
    // incompatibility that silently invalidated genuine signatures, and a
    // diagnostic nobody can read is how that reaches production unnoticed.
    return unverifiable(`CMS signature check failed -- ${describeError(error)}`);
  }
}
